pre_commit/languages/renv.py

```diff
@@ -0,0 +1,97 @@
+from pre_commit.envcontext import UNSET
+from pre_commit.hook import Hook
+from pre_commit.languages import helpers
+from pre_commit.prefix import Prefix
+from pre_commit.util import clean_path_on_failure
+from pre_commit.util import cmd_output_b
+
+ENVIRONMENT_DIR = 'renv'
+get_default_version = helpers.basic_get_default_version
+healthy = helpers.basic_healthy
+
+
+@contextlib.contextmanager
+def in_env(
+        prefix: Prefix,
+        language_version: str,
+) -> Generator[None, None, None]:
+    # probably not needed as no env have to be set.
+    # Remove this and get_env_patch altogether or keep
+    # for consistency with other languages?
+    envdir = prefix.prefix_dir
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+def get_env_patch(venv: str) -> PatchesT:
+    # Assumption: location of Rscript on PATH
+    return (('R_HOME', UNSET),)
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+    directory = helpers.environment_dir(ENVIRONMENT_DIR, version)
+    # installation location currently hardcoded
+    # https://github.com/rstudio/renv/issues/472
+    env_dir = prefix.path('renv')
+    # env_dir = prefix.path(directory)
+    os.mkdir(prefix.path(directory))  # directory needed in _write_state()
+    with clean_path_on_failure(env_dir):
+        cmd_output_b(
+            'Rscript', '--vanilla', '-e',
+            """
+            missing_pkgs <- setdiff(
+                "renv", unname(installed.packages()[, "Package"])
+            )
+            options(
+                repos = c(CRAN = "https://cloud.r-project.org"),
+                renv.consent = TRUE
+            )
+            install.packages(missing_pkgs)
+            if (file.exists('renv.lock')) {
+                renv::restore()
+            } else {
+                renv::init()
+            }
+            # TODO remove this workaround when renv > 0.12.5 is on CRAN and
+            # check for minimal version requirement:
+            # https://github.com/rstudio/renv/issues/634
+            install.packages("remotes")
+            remotes::install_local()
+            """,
+            cwd=prefix.prefix_dir,
+        )
+        if additional_dependencies:
+            cmd_output_b(
+                'Rscript', '-e',
+                """
+                renv::install(commandArgs(trailingOnly = TRUE))
+                """, *additional_dependencies,
+                cwd=prefix.prefix_dir,
+            )
+
+
+def run_hook(
+        hook: Hook,
+        file_args: Sequence[str],
+        color: bool,
+) -> Tuple[int, bytes]:
+    with in_env(hook.prefix, hook.language_version):
+        # activates renv through .Rprofile in
+        # hook.prefix.prefix_dir
+        # expose entry to the user for maximal control
+        return helpers.run_xargs(
+            hook, hook.cmd, file_args,
+            color=color, cwd=hook.prefix.prefix_dir,
```

Comment:
the current expectation for hook authors is that hooks will be executed from the root of the repository, so `cwd=...` won't be workable here

I think that necessarily changes how the environment setup is working now -- the ~normalish way of setting this up is by pointing some set of environment variables at the ENVIRONMENT_DIR (and subdirs)

note also that things must be installed entirely inside the ENVIRONMENT_DIR or they potentially won't be persisted by caching

Reply:
> the current expectation for hook authors is that hooks will be executed from the root of the repository, so cwd=... won't be workable here.

of course 🤦. 

>  the ~normalish way of setting this up is by pointing some set of environment variables at the ENVIRONMENT_DIR (and subdirs).

That's what I saw. Problem is (and that's a peculiarity of R) the most operations happen in an R session, not the terminal session. `renv` can only be activated properly from inside R. For that reason, my current suggestion is to place a `renv::activate()` call in a file, then point the `R_PROFILE_USER` environment variable to that file so the code will run at [R startup](https://rstats.wtf/r-startup.html). The idea is now to allow two versions, inline code expressions and file in `entry:`:
```bash
Rscript -e expr # like Rscript -e "1+1" 
Rscript path/to/file 
```
An inject the options `--no-save`, `--no-restore`, `--no-site-file` and `--no-environ` after `Rscript` (to basically have `--vanilla` start, but run this one `renv::activate()` statement) and expand the path similar to `language: script` since we have to run the hooks in the working directory of the repo. The files and the hook args are then passed after the above, i.e. 
```r
Rscript --no-save --no-restore --no-site-file --no-environ /abs/path/ [files] [hook args]
Rscript --no-save --no-restore --no-site-file --no-environ -e expr [files] [hook args]
```


> note also that things must be installed entirely inside the ENVIRONMENT_DIR or they potentially won't be persisted by caching.

This should be solved now. Just placing the whole env in a subdirectory now.